### PR TITLE
Optimise hot reading from remote tables

### PR DESCRIPTION
**Pull request by nickitat**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* for reading from remote tables we use smaller tasks (instead of reading the whole part) to make tasks stealing work
* task size is determined by size of columns to read
* always use 1MB buffers for reading from s3
* boundaries of cache segments aligned to 1MB so they have decent size even with small tasks. it also should prevent fragmentation

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
This is an automated comment for commit 979b68a488c88bad53a44e4c9a8c525572091de5 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49287/979b68a488c88bad53a44e4c9a8c525572091de5/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp [hunk 1]
```diff
@@ -49,7 +49,7 @@ CachedOnDiskReadBufferFromFile::CachedOnDiskReadBufferFromFile(
     bool allow_seeks_after_first_read_,
     bool use_external_buffer_,
     std::optional<size_t> read_until_position_)
-    : ReadBufferFromFileBase(settings_.remote_fs_buffer_size, nullptr, 0, file_size_)
+    : ReadBufferFromFileBase(use_external_buffer_ ? 0 : settings_.remote_fs_buffer_size, nullptr, 0, file_size_)
```
Comment (kssenii) [inline src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp:52]: could you please move this into a separate PR so that it is merged sooner? it is an important fix

File: src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp [hunk 2]
```diff
@@ -149,11 +149,8 @@ CachedOnDiskReadBufferFromFile::getCacheReadBuffer(const FileSegment & file_segm
     ReadSettings local_read_settings{settings};
     /// Do not allow to use asynchronous version of LocalFSReadMethod.
     local_read_settings.local_fs_method = LocalFSReadMethod::pread;
-
-    // The buffer will unnecessarily allocate a Memory of size local_fs_buffer_size, which will then
-    // most likely be unused because we're swap()ping our own internal_buffer into
-    // implementation_buffer before each read. But we can't just set local_fs_buffer_size = 0 here
-    // because some buffer implementations actually use that memory (e.g. for prefetching).
+    if (use_external_buffer)
+        local_read_settings.local_fs_buffer_size = 0;
```
Comment (kssenii) [inline src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp:153]: and this one

File: src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp [hunk 3]
```diff
@@ -815,6 +817,9 @@ bool CachedOnDiskReadBufferFromFile::nextImplStep()
                 }
             }
 
+            if (use_external_buffer && !internal_buffer.empty())
+                internal_buffer.resize(original_buffer_size);
```
Comment (kssenii) [inline src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp:821]: I do not understand this change, could you please explain why this is needed?

Comment (nickitat) [reply src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp:821]: we did resize on l.852 and after reading is done we need to roll it back

File: src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp (hunk 1 above)
Comment (nickitat) [reply src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp:52]: ok

File: src/Storages/MergeTree/MergeTreePrefetchedReadPool.cpp [hunk 4]
```diff
@@ -366,6 +366,9 @@ MergeTreePrefetchedReadPool::PartsInfos MergeTreePrefetchedReadPool::getPartsInf
         part_info->column_name_set = {required_column_names.begin(), required_column_names.end()};
         part_info->task_columns = task_columns;
 
+        if (settings.prefetch_buffer_size < DBMS_DEFAULT_BUFFER_SIZE)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "remove me");
+
```
Comment (nickitat) [inline src/Storages/MergeTree/MergeTreePrefetchedReadPool.cpp:371]:
```suggestion
```

File: src/Storages/MergeTree/MergeTreeReadPool.cpp [hunk 5]
```diff
@@ -44,13 +52,40 @@ MergeTreeReadPool::MergeTreeReadPool(
     , parts_ranges(std::move(parts_))
     , predict_block_size_bytes(context_->getSettingsRef().preferred_block_size_bytes > 0)
     , do_not_steal_tasks(do_not_steal_tasks_)
+    , merge_tree_use_const_size_tasks_for_remote_reading(context_->getSettingsRef().merge_tree_use_const_size_tasks_for_remote_reading)
     , backoff_settings{context_->getSettingsRef()}
     , backoff_state{threads_}
 {
+    if (std::ranges::count(is_part_on_remote_disk, true))
```
Comment (kssenii) [inline src/Storages/MergeTree/MergeTreeReadPool.cpp:59]: hm, `is_part_on_remote_disk` is initialized in `fillPerPartInfo` below this check, so this check is always false?

File: src/Interpreters/Cache/FileCache_fwd.h [hunk 6]
```diff
@@ -4,7 +4,7 @@
 namespace DB
 {
 
-static constexpr int FILECACHE_DEFAULT_MAX_FILE_SEGMENT_SIZE = 100 * 1024 * 1024;
+static constexpr int FILECACHE_DEFAULT_MAX_FILE_SEGMENT_SIZE = 8 * 1024 * 1024;
```
Comment (kssenii) [inline src/Interpreters/Cache/FileCache_fwd.h:7]: isn't it too small?

File: src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp [hunk 7]
```diff
@@ -840,6 +848,11 @@ bool CachedOnDiskReadBufferFromFile::nextImplStep()
 
     chassert(!internal_buffer.empty());
 
+    /// We allocate buffers not less than 1M so that s3 requests will not be too small. But the same buffers (members of AsynchronousReadIndirectBufferFromRemoteFS)
+    /// are used for reading from files. Some of these readings are fairly small and their performance degrade when we use big buffers (up to ~20% for queries like Q23 from ClickBench).
+    if (use_external_buffer && read_type == ReadType::CACHED && settings.local_fs_buffer_size < internal_buffer.size())
+        internal_buffer.resize(settings.local_fs_buffer_size);
```
Comment (kssenii) [inline src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp:854]: below when calling predownload() we might decide to bypass cache and change read_method, so we need to resize the buffer back somewhere there

File: src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp [hunk 8]
```diff
@@ -42,21 +42,25 @@ namespace ErrorCodes
     extern const int ARGUMENT_OUT_OF_BOUND;
 }
 
+static size_t chooseBufferSize(const ReadSettings & settings, size_t file_size)
+{
+    /// Buffers used for prefetch or pre-download better to have enough size, but not bigger than the whole file.
+    return std::min<size_t>(std::max<size_t>(settings.prefetch_buffer_size, DBMS_DEFAULT_BUFFER_SIZE), file_size);
```
Comment (kssenii) [inline src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp:48]: is max with DBMS_DEFAULT_BUFFER_SIZE necessary? seems redundant, if user makes the buffer too small it's his fault? as by default prefetch_buffer_size is equal to DBMS_DEFAULT_BUFFER_SIZE then it should be ok

Comment (nickitat) [reply src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp:48]: https://github.com/ClickHouse/ClickHouse/blob/3dfc0bd26597ca8c9e6c30343470de15553f3ac3/src/Storages/MergeTree/MergeTreeReaderStream.cpp#L87
we adjust buffer size (including prefetch buffer) based on mark range sizes. they could be fairly small (few KBs).
since we use them to download segments that (almost) cannot be smaller than 1MB I think it makes sense to enforce this lower bound.

File: src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp (hunk 7 above)
Comment (nickitat) [reply src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp:854]: if I'm not missing smth, we do resize only if `read_type == ReadType::CACHED` and in the only place where we assign positive value to `bytes_to_predownload` we also set `read_type = REMOTE_FS_READ_AND_PUT_IN_CACHE`

File: src/Interpreters/Cache/FileCache_fwd.h (hunk 6 above)
Comment (nickitat) [reply src/Interpreters/Cache/FileCache_fwd.h:7]: first of all, why change it. since we start to share segments between tasks ('cause of alignment) now some of them that actually need to read only from the end of the segment could read the whole segment in pre-download. so it seems logical to limit size of the segments.
from what perspective do you think it might be too small? overhead per segment during processing seems very small. the only downside I could think of is prob number of GetObject-s will increase.

File: src/Storages/MergeTree/MergeTreeReadPool.cpp (hunk 5 above)
Comment (nickitat) [reply src/Storages/MergeTree/MergeTreeReadPool.cpp:66]: what a shame )

File: src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp (hunk 7 above)
Comment (kssenii) [reply src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp:854]: ah, yes, sorry

File: src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp (hunk 8 above)
Comment (kssenii) [reply src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp:48]: yes, good point

File: src/Interpreters/Cache/FileCache_fwd.h (hunk 6 above)
Comment (nickitat) [reply src/Interpreters/Cache/FileCache_fwd.h:7]: I also saw this effect, in my case it was because of connection timeouts that start to happen more frequently when we increase number of parallel requests. for this i added conn pool for s3 reading in the second pr.
and I like your idea of coalescing reads on GetObject level

**Review by kssenii — APPROVED** [review]

File: src/Interpreters/executeQuery.cpp [hunk 9]
```diff
@@ -322,6 +322,8 @@ static std::tuple<ASTPtr, BlockIO> executeQueryImpl(
     /// This does not have impact on the final span logs, because these internal queries are issued by external queries,
     /// we still have enough span logs for the execution of external queries.
     std::shared_ptr<OpenTelemetry::SpanHolder> query_span = internal ? nullptr : std::make_shared<OpenTelemetry::SpanHolder>("query");
+    if (query_span)
+        LOG_DEBUG(&Poco::Logger::get("executeQuery"), "Query span trace_id for opentelemetry log: {}", query_span->trace_id);
```
Comment (azat) [inline src/Interpreters/executeQuery.cpp:326]: Maybe it worth to check for `query_span->isTraceEnabled()` ? To print trace_id only if tracing is enabled

File: src/Storages/MergeTree/MergeTreeReadPool.cpp [hunk 10]
```diff
@@ -300,6 +332,8 @@ void MergeTreeReadPool::fillPerThreadInfo(
             parts_queue.push(std::move(info.second));
     }
 
+    LOG_DEBUG(log, "min_marks_for_concurrent_read={}", min_marks_for_concurrent_read);
```
Comment (azat) [inline src/Storages/MergeTree/MergeTreeReadPool.cpp:335]: This looks like it deserves to be under `LOG_TEST`
